hedera-node/docs/design/services/token-service/airdrops/airdrop-transaction.md

```diff
@@ -0,0 +1,206 @@
+ * - Each pending transfer _updated_ SHALL be added to the `pending_airdrops` field for the record.
+ */
+ message TokenAirdropTransactionBody {
+/**
+     * A list of token transfers representing one or more airdrops.<br/>
+     * The sender for each transfer MUST have sufficient balance to complete the transfers.
+     * <p>
+     * All token transfers MUST successfully transfer tokens or create a pending airdrop
+     * for this transaction to succeed.<br/>
+     * This list MUST contain between 1 and 10 transfers, inclusive.
+     * <p>
+     * <blockquote>Note that each transfer of fungible/common tokens requires both a debit and
+     * a credit, so each _fungible_ token transfer MUST have _balanced_ entries in the
+     * TokenTransferList for that transfer.</blockquote>
+     */
+     repeated TokenTransferList token_transfers= 1;
+}
+```
+
+Add new RPC to `TokenService` :
+
+```protobuf
+   /**
+     * Airdrop one or more tokens to one or more accounts.<br/>
+     * This distributes tokens from the balance of one or more sending account(s) to the balance
+     * of one or more recipient accounts. Accounts will receive the tokens in one of four ways.
+     * <ul>
+     *   <li>An account already associated to the token to be distributed SHALL receive the
+     *       airdropped tokens immediately to the recipient account balance.</li>
+     *   <li>An account with available automatic association slots SHALL be automatically
+     *       associated to the token, and SHALL immediately receive the airdropped tokens to the
+     *       recipient account balance.</li>
+     *   <li>An account with "receiver signature required" set SHALL have a "Pending Airdrop"
+     *       created and MUST claim that airdrop with a `claimAirdrop` transaction.</li>
+     *   <li>An account with no available automatic association slots SHALL have a
+     *       "Pending Airdrop" created and MUST claim that airdrop with a `claimAirdrop`
+     *       transaction. </li>
+     * </ul>
+     * Any airdrop that completes immediately SHALL be irreversible. Any airdrop that results in a
+     * "Pending Airdrop" MAY be canceled via a `cancelAirdrop` transaction.<br/>
+     * All transfer fees (including custom fees and royalties), as well as the rent cost for the
+     * first auto-renewal period for any automatic-association slot occupied by the airdropped
+     * tokens, SHALL be charged to the account submitting this transaction.
+     */
+    rpc airdropTokens (Transaction) returns (TransactionResponse);
+```
+
+### Fee types
+
+The airdrop transaction would include several types of fees, a different set of them would be applicable, based on the conditions the airdrop is performed to. The list of fees is:
+
+- Airdrop fee covering the operation itself
+- Association fee covering a pre-paid fee for associating the token being airdropped to the receiver, in case there is a missing relationship between the two
```

Comment:
This is only if the `maxAutoAssociations` is not greater than `usedAutoAssociations` right?

Reply:
This association fee is not related to auto associations, but for explicit association which would happen in case the airdropped is claimed by the receiver. Since we don't want to add this expense to the receiving side, the airdrop sender should pay in advance for this potential future association.

If there are free auto association slots for the recipient, the airdrop would directly be transferred to the recipient and I think no association fee should be paid in this case.

Reply:
Based on the HIP the sender will pay for the automatic association for a recipient who has an automatic association slot available.
>As an hts-token-airdrop-sender I want to cover the fees (custom, association & rent until the end of the first full auto renewal period) of an airdrop that consumes an automatic token slot to reduce the friction for a receiving account

Reply:
Oh, I thought automatic associations are not covered by fees and only a transfer fee is paid in those cases. If they are, this association fee would also cover this auto association step, only if `maxAutoAssociations` >= `usedAutoAssociations`, so there are free auto association slots available.

Reply:
Interesting, I had understood that automatic association slots had a cost (both a fee when consumed, and a "rent" when the account is renewed).

Thanks for clarifying.

Reply:
Actually in the diagram in the HIP we can see that even if the receiver has free auto associate slots, the airdrop sender should still cover an association fee, so I think @jsync-swirlds is right and the association fee charged during an airdrop covers both a direct auto association or a potential manually triggered association, during a claim by the recipient. In any case, we should charge this association fee from the airdrop sender. That's why we also change the existing `CryptoTransfer` fee logic by making the sender always paying for the token association.

@Nana-EC , @Neeharika-Sompalli , what do you think?

<img width="1297" alt="image" src="https://github.com/hashgraph/hedera-services/assets/5901140/0a07bdb8-fa2c-42dd-91a6-3863cfe1c3ef">

Reply:
Ah I see .. so, we don't charge `maxAutoAssociations` during creation anymore. Then this makes sense.